### PR TITLE
Warn users that uses length to check if list is not empty on guards

### PR DESCRIPTION
**Pull request by ulissesalmeida**
Some Elixir users check if the list is not
empty by using the length function. It's not the optimal solution since it has to transverse the whole
list. Now we're warning the users and suggesting them
to use `!= []`.

Apply https://github.com/elixir-lang/elixir/issues/7513 enhancement.

### CODE REVIEW

File: lib/elixir/src/elixir_clauses.erl [hunk 1]
```diff
@@ -54,6 +54,13 @@ warn_zero_length_guard({{'.', _, [erlang, '==']}, Meta,
                           "Prefer to pattern match on an empty list or use "
                           "\"~ts == []\" as a guard", [ArgString, ArgString]),
   elixir_errors:warn(?line(Meta), ?key(E, file), Message);
+warn_zero_length_guard({{'.', _, [erlang, '>']}, Meta,
+                        [{{'.', _, [erlang, length]}, _, [Arg]}, 0]}, E) ->
+  ArgString = 'Elixir.Macro':to_string(Arg),
+  Message = io_lib:format("\"length(~ts) > 0\" is discouraged since it has to "
+                          "traverse the whole list to check if it is not empty. "
+                          "Prefer to use \"~ts != []\" as a guard", [ArgString, ArgString]),
```
Comment (josevalim) [inline lib/elixir/src/elixir_clauses.erl]: What do you think about the last sentence being:

> Prefer to pattern match on a cons cell, such as [_ | _], or use \"~ts != []\" as a guard

Comment (josevalim) [reply lib/elixir/src/elixir_clauses.erl]: Note: I changed my comment above to say "non-empty list" instead of "cons cell"

**Comment by josevalim** [comment]
@ulissesalmeida please run `bin/elixir bin/mix format` at the root of the project as well to format the warning_test file. :)

**Comment by brodeuralexis** [comment]
Why not modify `Kernel.length/1` to pattern match ?
```ex
def length([]), do: 0
```

**Comment by ulissesalmeida** [comment]
The problem is when the list is not empty. We need to transverse the entire
list to know its size

Em Qui, 29 de mar de 2018 08:05, Alexis Brodeur <notifications@github.com>
escreveu:

File: lib/elixir/src/elixir_clauses.erl [hunk 2]
```diff
@@ -54,6 +54,13 @@ warn_zero_length_guard({{'.', _, [erlang, '==']}, Meta,
                           "Prefer to pattern match on an empty list or use "
                           "\"~ts == []\" as a guard", [ArgString, ArgString]),
   elixir_errors:warn(?line(Meta), ?key(E, file), Message);
+warn_zero_length_guard({{'.', _, [erlang, '>']}, Meta,
+                        [{{'.', _, [erlang, length]}, _, [Arg]}, 0]}, E) ->
+  ArgString = 'Elixir.Macro':to_string(Arg),
+  Message = io_lib:format("\"length(~ts) > 0\" is discouraged since it has to "
```
Comment (eksperimental) [inline lib/elixir/src/elixir_clauses.erl]: I would change the meaning of the sentence in an affirmative way:
> Is discouraged since it has to transverse the whole list just to see if it's empty

File: lib/elixir/src/elixir_clauses.erl [hunk 3]
```diff
@@ -52,7 +52,15 @@ warn_zero_length_guard({{'.', _, [erlang, '==']}, Meta,
...
                           "traverse the whole list to check if it is empty or not. "
                           "Prefer to pattern match on an empty list or use "
-                          "\"~ts == []\" as a guard", [ArgString, ArgString]),
+                          "\"~ts == []\" as a guard.", [ArgString, ArgString]),
+  elixir_errors:warn(?line(Meta), ?key(E, file), Message);
+warn_zero_length_guard({{'.', _, [erlang, '>']}, Meta,
+                        [{{'.', _, [erlang, length]}, _, [Arg]}, 0]}, E) ->
+  ArgString = 'Elixir.Macro':to_string(Arg),
+  Message = io_lib:format("\"length(~ts) > 0\" is discouraged since it has to "
+                          "traverse the whole list to check if it is empty or not. "
```
Comment (eksperimental) [inline lib/elixir/src/elixir_clauses.erl]: I think just mentioning that the list is empty is enough. I find the "or not" part unnecessary, specially for an error msg that aims to be as concise as possible

Comment (eksperimental) [reply lib/elixir/src/elixir_clauses.erl]: Additionally I would rephrase my previous suggestion since it is not clear what is the "it" in the "it has to traverse...". I wouldngo with this:
> is discouraged since the whole list needs to be traversed in order to check if it's empty

**Comment by josevalim** [comment]
Let’s keep the message as is because it is consistent with the existing
one. I will do another pass after merging incorporating suggestions. :)
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/elixir/src/elixir_clauses.erl (hunk 3 above)
Comment (josevalim) [reply lib/elixir/src/elixir_clauses.erl]: I think we should avoid passive voice in general as we just moved the ambiguity elsewhere. Previously it was unclear what was the "it", now it is unclear who is causing the action altogether.
